src/database/models/userArticleLink.json

```diff
@@ -0,0 +1,56 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "cofacts/schemas/userArticleLink",
+  "$comment": "design document https://g0v.hackmd.io/eIeU2g86Tfu5VnLazNfUvQ",
+  "description": "A representation of an UserArticleLink",
+  "title": "UserArticleLink",
+  "type": "object",
+
+  "properties": {
+    "id": {
+      "isObjectId": true
+    },
+    "userId": {
+      "type": "string"
+    },
+    "articleId": {
+      "type": "string"
+    },
+    "createdAt": {
+      "format": "date-time",
+      "instanceOf": "Date"
+    },
+    "lastViewedAt": {
+      "oneOf": [ {
+        "type": "null"
+      }, {
```

Comment:
But `null` and `undefined` are different in mongodb. For example 

```
db.collection('inventory').find({
  item: null
});
```

```
db.collection('inventory').find({
  item: { $exists: false }
});
```

I'm not sure there is any effect in the future, so I choose the precise defines.

Reply:
Yeah I know about the `$exists`. But can you set an optional field to `null` if you say that the field is some non-null type in the validator?

Reply:
Yes, if using mongodb

```
$> db.ppl.insert({phoneNumber: null})
WriteResult({ "nInserted" : 1 })
$> db.ppl.insert({phoneNumber: '0912'})
WriteResult({ "nInserted" : 1 })
$> db.ppl.find()
{ "_id" : ObjectId("5eb3ad6a9a5e1f8e59885998"), "phoneNumber" : null }
{ "_id" : ObjectId("5eb3ad739a5e1f8e59885999"), "phoneNumber" : "0912" }
```

There are 2 options: 
1. [require=true] Should be exist, then string or null
2. [require=false] It would be string or undefined (not exists)

But I think yours is better to be less verbose first, and then update if need.

Reply:
btw, If we follow option 1, then we can query easily
```
db.collection('inventory').find({
  itemA: null,
  itemB: 'book'
});
```

Not sure if it's important for now

Reply:
Hmm if I want to find a document that has one field that does not exist, I would use `{itemA: {$exists: false}}`. It also mixes well with other conditions in place (i.e. `{itemA: {$exists: false, itemB: 'book'}}`). Slightly longer than the `null` version, but I think its occurrence in codebase should be less than field definitions.

Reply:
Seems that json schema we are currently using do blocks non-required fields from being set to `null`.
![image](https://user-images.githubusercontent.com/108608/81265339-824ed800-9075-11ea-8365-a52860f545f6.png)

Reply:
On the other hand, one draw back (?) of using non-required fields would be that it's a bit verbose if we want to set a field conditionally, like:
```js
const person = {
  name: 'foo',
  ...(revealGender ? {gender: input} : {})
}
```

Setting everything to required force us to specify all fields, but is easier (?) to set conditional content
```js
const person = {
  name: 'foo',
  gender: revealGender ? input : null,
  bloodType: null,
  weight: null,
  height: null,
  birthday: null
}
```

Reply:
> Seems that json schema we are currently using _does_ block non-required fields from being set to `null`.

You are not using the current one ? Should be use `oneOf` in your demo case ?

Reply:
> Hmm if I want to find a document that has one field that does not exist, the first idea that comes to my mind would be `{itemA: {$exists: false}}`, did not think of testing `null` in the first place.

I'm not sure that is the first idea. There are different usages in different situations. Let's talk about `newReplyNotifyToken` if it's expired or not working or something. I want to set to be `null`, not `undefined`, and find those documents by using `null` not `{$exists: false}`

Reply:
In the demo case I am testing if jsonschema can block `null` when it's assigned to non-required fields.

Personally I think we should either stick to (a) field not-required + can't be null, or (b) all-field-required + can be null, not a mixture of (a) and (b). While the current code is in (b), in the demonstration I was trying to show that (a) also works as exptected.

Reply:
Interesting aspect of `newReplyNotifyToken`, didn't think of that. When `newReplyNotifyToken` is expired, I would ask the user if they want to delete the token. If they say yes, we invoke `db.collection('userArticleLinks').update({$unset: {newReplyNotifyToken: ''}})`. That's indeed longer than setting `newReplyNotifyToken` to `null`.